File: code/872.leaf-similar-trees.php
<?php
/*
 * @lc app=leetcode id=872 lang=php
 *
 * [872] Leaf-Similar Trees
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root1
     * @param TreeNode $root2
     * @return Boolean
     */
    function leafSimilar($root1, $root2) {
        $leafs1 = [];
        $leafs2 = [];
        $this->searchLeaf($root1, $leafs1);
        $this->searchLeaf($root2, $leafs2);
        return $leafs1 == $leafs2;
    }

    function searchLeaf($root, &$leafs) {
        if ($root->left != null) {
            $this->searchLeaf($root->left, $leafs);
        }

        if ($root->right != null) {
            $this->searchLeaf($root->right, $leafs);
        }

        if ($root->left == null && $root->right == null) {
            $leafs[] = $root->val;
        }
    }
}
// @lc code=end

File: code/872.leaf-similar-trees.go
package code

import (
	"slices"
)

/*
 * @lc app=leetcode id=872 lang=golang
 *
 * [872] Leaf-Similar Trees
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	leaves1 := make([]int, 0)
	leaves2 := make([]int, 0)

	recursion872(root1, &leaves1)
	recursion872(root2, &leaves2)

	return slices.Equal(leaves1, leaves2)
}

func recursion872(node *TreeNode, leaves *[]int) {
	if node.Left != nil {
		recursion872(node.Left, leaves)
	}

	if node.Right != nil {
		recursion872(node.Right, leaves)
	}

	if node.Left == nil && node.Right == nil {
		*leaves = append(*leaves, node.Val)
	}
	return
}

// @lc code=end
